packages/validation/src/references/process.ts

```diff
@@ -0,0 +1,110 @@
+import groupBy from "lodash/groupBy.js";
+import keyBy from "lodash/keyBy.js";
+import uniq from "lodash/uniq.js";
+import { ResourceParser } from "../common/resourceParser.js";
+import {
+  Incremental,
+  Resource,
+  ResourceRefsProcessingConfig,
+} from "../common/types.js";
+import { handlePairRefMapping } from "./handlePairRefMapping.js";
+import { handleRefMappingByKey } from "./handleRefMappingByKey.js";
+import { getOutgoingRefMappers } from "./mappers/index.js";
+import { cleanResourceRefs } from "./utils/cleanResourceRefs.js";
+import { clearOutgoingResourceRefs } from "./utils/clearOutgoingRefs.js";
+import { getResourceKindsWithTargetingRefs } from "./utils/getResourceKindsWithTargetingRefs.js";
+import { getResourceRefNodes } from "./utils/getResourceNodes.js";
+import { refMapperMatchesKind } from "./utils/helpers.js";
+
+/**
+ * Processes resources and MUTATES them with their references to other resources.
+ *
+ * @glossary
+ * - RefMap: Decides whether an edge is formed between two resources.
+ * - Sibling: A RefMap is build on a single property that will be higlighted,
+ *     though target identity might be based on multiple properties. These
+ *     non-highlighted properties are called siblings.
+ */
+export function processRefs(
+  resources: Resource[],
+  parser: ResourceParser,
+  incremental?: Incremental
+) {
+  const filteredResources = filterResources(resources, incremental);
+
+  return doProcessRefs(resources, filteredResources, {
+    shouldIgnoreOptionalUnsatisfiedRefs: false,
+    parser,
```

Comment:
Currently I'm passing config with parser throughout the code as some sort of context object. It could use a refactor in the future but for now it'll do.